[RISCV][GISel] Add instruction selection for G_SEXT, G_ZEXT, and G_SEXT_INREG

mshockwave opened the pull request:
G_SEXT and G_ZEXT are supported via patterns imported from SDISel; G_SEXT_INREG is selected using hand-written code as there is no (functional) rule at this moment to import G_SEXT_INREG from ISD::SEXT_INREG.

Credit helps from @topperc on G_SEXT and G_ZEXT.

mshockwave committed:
[RISCV][GISel] Add instruction selection for G_SEXT, G_ZEXT, and G_SEXT_INREG

G_SEXT and G_ZEXT are supported via patterns imported from SDISel;
G_SEXT_INREG is selected using hand-written code as there is no
(functional) rule at this moment to import G_SEXT_INREG from sext_inreg.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.td:
```diff
@@ -1774,6 +1774,12 @@ let Predicates = [IsRV64], hasSideEffects = 0, mayLoad = 0, mayStore = 0,
 def PseudoZEXT_W : Pseudo<(outs GPR:$rd), (ins GPR:$rs), [], "zext.w", "$rd, $rs">;
 } // Predicates = [IsRV64], ...
 
+let Predicates = [IsRV64] in
```
Add the new patterns to RISCVGISel.td for now since they aren't needed by SelectionDAG.,

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -363,6 +366,32 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
+  const MachineOperand &Dst = MI.getOperand(0);
+  assert(Src.isReg() && Dst.isReg());
+  // addiw rd, rs, 0 (i.e. sext.w rd, rs)
+  MachineInstr *NewMI = MIB.buildInstr(RISCV::ADDIW)
```
I think you can do `MIB.buildInstr(RISCV::ADDIW, {MI.getOperand(0)},  {MI.getOperand(1)}).addImm(0U);` This should automatically call getReg on the operand. I only learned this last week.

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -363,6 +366,32 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
+
+  const MachineOperand &Src = MI.getOperand(1);
+  const MachineOperand &Dst = MI.getOperand(0);
+  assert(Src.isReg() && Dst.isReg());
```
You can drop this assert. getReg() will assert if they aren't registers.

mshockwave committed:
Addressed comments

 - Moved newly created isel patterns into RISCVGISel.td
 - Some minor fixes in `selectSExtInreg`

topperc reviewed (approved):
LGTM
